runtime/nvqir/NVQIR.cpp

```diff
@@ -135,6 +136,55 @@ std::size_t qubitToSizeT(Qubit *q) {
   return q->idx;
 }
 
+template <typename T>
+constexpr std::string_view typeName() {
+  static_assert(
+      std::is_same<T, float>::value || std::is_same<T, double>::value,
+      "unsupported type for array conversion to simulation precision");
+  return std::is_same<T, float>::value ? "float" : "double";
+}
+
+/// Input was complex<float>/complex<double> but we prefer
+/// complex<double>/complex<float>. Make a copy, extending or truncating the
+/// values.
+template <typename To, typename From>
+std::unique_ptr<std::complex<To>[]> convertToComplex(std::complex<From> *data,
+                                                     std::size_t size) {
+  auto stateSize = pow(2, size);
+  constexpr auto toType = typeName<To>();
+  constexpr auto fromType = typeName<From>();
+
+  cudaq::info("copying {} complex<{}> values to complex<{}>", stateSize,
+              fromType, toType);
+
+  auto convertData = std::make_unique<std::complex<To>[]>(stateSize);
+  for (uint64_t i = 0; i < stateSize; ++i)
+    convertData[i] = std::complex<To>{static_cast<To>(data[i].real()),
+                                      static_cast<To>(data[i].imag())};
+  return convertData;
+}
+
+/// Input was complex<float>/complex<double> but we prefer
+/// complex<double>/complex<float>. Make a copy, extending or truncating the
+/// values.
+template <typename To, typename From>
```

Comment:
Can we use concepts here to ensure that `To` and `From` are both in the set `{ float, double }`?

Reply:
Done!